GPflowOpt/transforms.py

```diff
@@ -93,38 +80,74 @@ def __init__(self, A, b):
         A is not invertible, hence inverse and backward are not supported.
         :param b: A P-dimensional offset vector.
         """
-        self.b = np.atleast_1d(b)
-        self.A = np.atleast_1d(A)
-        if len(self.A.shape) == 1:
-            self.A = np.diag(self.A)
+        super(LinearTransform, self).__init__()
+        assert A is not None
+        assert b is not None
 
-        assert (len(self.b.shape) == 1)
-        assert (len(self.A.shape) == 2)
+        b = np.atleast_1d(b)
+        A = np.atleast_1d(A)
+        if len(A.shape) == 1:
+            A = np.diag(A)
 
-    def forward(self, X):
-        return np.dot(np.atleast_2d(X), self.A.T) + self.b
+        assert (len(b.shape) == 1)
+        assert (len(A.shape) == 2)
 
+        self.A = DataHolder(A)
+        self.b = DataHolder(b)
+
+    def build_forward(self, X):
+        return tf.matmul(X, tf.transpose(self.A)) + self.b
+
+    @AutoFlow((float_type, [None, None]))
     def backward(self, Y):
         """
-        Overwrites the default backward approach, it avoids an explicit matrix inversion.
+        Overwrites the default backward approach, to avoids an explicit matrix inversion.
         """
-        L = scipy.linalg.cholesky(self.A.T)
-        return scipy.linalg.cho_solve((L, False), (Y - self.b).T).T
-
-    def build_forward(self, X):
-        return tf.matmul(X, self.A.T) + self.b
+        return self.build_backward(Y)
 
     def build_backward(self, Y):
         """
-        Overwrites the default backward approach, it avoids an explicit matrix inversion.
+        TensorFlow implementation of the inverse mapping
         """
-        L = tf.cholesky(self.A.T)
-        XT = tf.matrix_triangular_solve(tf.transpose(L), tf.matrix_triangular_solve(L, tf.transpose(Y - self.b)))
+        L = tf.cholesky(tf.transpose(self.A))
+        XT = tf.cholesky_solve(L, tf.transpose(Y-self.b))
         return tf.transpose(XT)
 
+    def build_backward_variance(self, Yvar):
+        """
+        Additional method for scaling variance backward (used in :class:`.Normalizer`). Can process both the diagonal
+        variances returned by predict_f, as well as full covariance matrices.
+        :param Yvar: N x N x P or N x P
+        :return: Yvar scaled, same rank and dimensionality as input
+        """
+        rank = tf.rank(Yvar)
```

Comment:
does Yvar has a static TF shape? Would make things easier.

As a sidenote, for prediction that is harder to do but I wish GPFlow would do that for like training data where the shape is known upfront in DataHolder.